src/charm.py

```diff
@@ -804,120 +803,47 @@ def _set_leader_unit_active_if_needed(self):
-        return False
-
-    def _juju_secret_get_key(self, scope: Scopes, key: str) -> Optional[str]:
-        if not key:
-            return
-
-        if self._juju_secrets_get(scope):
-            secret_cache = self.secrets[scope].get(Config.Secrets.SECRET_CACHE_LABEL)
-            if secret_cache:
-                secret_data = secret_cache.get(key)
-                if secret_data and secret_data != Config.Secrets.SECRET_DELETED_LABEL:
-                    logging.debug(f"Getting secret {scope}:{key}")
-                    return secret_data
-        logging.debug(f"No value found for secret {scope}:{key}")
-
     def get_secret(self, scope: Scopes, key: str) -> Optional[str]:
         """Getting a secret."""
-        return self._juju_secret_get_key(scope, key)
-
-    def _juju_secret_set(self, scope: Scopes, key: str, value: str) -> str:
-        """Helper function setting Juju secret."""
-        peer_data = self._peer_data(scope)
-        self._juju_secrets_get(scope)
-
-        secret = self.secrets[scope].get(Config.Secrets.SECRET_LABEL)
-
-        # It's not the first secret for the scope, we can re-use the existing one
-        # that was fetched in the previous call
-        if secret:
-            secret_cache = self.secrets[scope][Config.Secrets.SECRET_CACHE_LABEL]
-
-            if secret_cache.get(key) == value:
-                logging.debug(f"Key {scope}:{key} has this value defined already")
-            else:
-                secret_cache[key] = value
-                try:
-                    secret.set_content(secret_cache)
-                except OSError as error:
-                    logging.error(
-                        f"Error in attempt to set {scope}:{key}. "
-                        f"Existing keys were: {list(secret_cache.keys())}. {error}"
-                    )
-                logging.debug(f"Secret {scope}:{key} was {key} set")
-
-        # We need to create a brand-new secret for this scope
-        else:
-            scope_obj = self._scope_opj(scope)
-
-            secret = scope_obj.add_secret({key: value})
-            if not secret:
-                raise SecretNotAddedError(f"Couldn't set secret {scope}:{key}")
-
-            self.secrets[scope][Config.Secrets.SECRET_LABEL] = secret
-            self.secrets[scope][Config.Secrets.SECRET_CACHE_LABEL] = {key: value}
-            logging.debug(f"Secret {scope}:{key} published (as first). ID: {secret.id}")
-            peer_data.update({Config.Secrets.SECRET_INTERNAL_LABEL: secret.id})
+        label = generate_secret_label(self, scope)
+        secret = self.secrets.get(label)
+        if not secret:
+            return
 
-        return self.secrets[scope][Config.Secrets.SECRET_LABEL].id
+        value = secret.get_content().get(key)
+        if value != Config.Secrets.SECRET_DELETED_LABEL:
+            return value
 
     def set_secret(self, scope: Scopes, key: str, value: Optional[str]) -> Optional[str]:
         """(Re)defining a secret."""
```

Comment:
please forgive my noob question, but why isn't it necessary to check if the unit is a leader

Reply:
As this question is coming up a lot, I thought to bring up a live example :-)

https://github.com/canonical/mongodb-k8s-operator/blob/main/lib/charms/mongodb/v0/mongodb_tls.py#L177

Typically TLS certs exist both on the `app` level (controlled by the leader unit) and on an individual level for units themselves. Thus units must be able to set their own secrets -- without being a leader.